omnibus/partybus/bin/partybus

```diff
@@ -43,9 +44,97 @@ def load_migrations
 end
 
 def set_initial_migration_state(migrations)
-  latest_migration = migrations.last
+  write_migration_state(migrations.last)
+end
+
+def write_migration_state(migration)
   File.open(Partybus.config.migration_state_file, 'w') do |f|
-      f.puts({:major => latest_migration.major, :minor => latest_migration.minor}.to_json)
+    f.puts({:major => migration.major, :minor => migration.minor}.to_json)
+  end
+end
+
+def migration_state_has_content?
+  !File.zero?(Partybus.config.migration_state_file)
+end
+
+#
+# We infer the migration state to try to repair installations that
+# have lost their migration state file. Since the data may have come
+# from a backup, we can not rely on the currently installed version to
+# be up-to-date and we don't necessarily know the previously installed
+# version.
+#
+# We know we can not support upgrades from further back than
+# Enterprise Chef 11.1, which shipped migration 1.13:
+#
+#   https://github.com/chef/chef-server/tree/11.1.0/files/private-chef-upgrades/001
+#
+# We also know that migration 1.14 through 1.19 landed together in
+# Chef Server 12.0.
+#
+# Luckily for us, migration 1.20 shipped a sqitch tag.  Further luck:
+# migrations above 1.20 generally appear to be safe-to-reapply so we
+# don't have to detect perfectly.
+#
+INFER_LOWER_BOUND_MAJOR=1
+INFER_LOWER_BOUND_MINOR=20
+def infer_migration_state(migrations, force)
+  if migration_state_has_content? && !force
+    log("Migration state file non-empty, inference not required.")
+    exit(0)
+  end
+
+  log("Infering migration-level from system state")
+  applied_version = nil
+  migrations.sort.reverse.each do |m|
+    if m.run_check
+      log "#{m}: ALREADY APPLIED"
+      applied_version = m
+      break
+    else
+      log "#{m}: NOT APPLIED"
+    end
+
+    # A number of migrations before 1.20 (the Chef Server 12.0) are
+    # mover migrations that are harder to check for.  If we make it
+    # all the way to here, we stop checking and hope one of the two
+    # special cases below apply.
```

Comment:
Looking to see if couch is still running/installed might give us a more definitive answer for minimum level that has or hasn't been applied.

Reply:
Unfortunately, I'm not sure it does because maybe they didn't run cleanup.  I suppose if we /don't/ see couchdb then we have more evidence that we go through migrations 14-19 (right now we check if migration 16 is completed and use that to assume that they successfully upgraded to 12.0.0).  If we /do/ see couchdb and we assume they would have run cleanup, then we have evidence that they haven't upgraded to 12, but I am not sure if I trust that more or less than just looking at migration 16.

I think I'm OK with this for now because it makes mostly conservative judgements. We are taking the following leaps of faith: (1) My reading of the idempotency of some of the mover migrations and (2) that most likely we aren't dealing with a half-upgraded Chef Server 12.0.0 installation.  I'm not sure looking at couchdb changes that substantially unlesss I'm missing something.